[![バナー画像](http://ai999.careers/bnr_jdla.png)](http://study-ai.com/jdla/)

URL: <http://study-ai.com/jdla/>

# 応用数学

## 第一章　線形代数

<b>要点まとめ</b>

<b>機械学習の分野では、与えられたデータの基本的性質を見出す処理として、分散・共分散行列に対する固有値・固有ベクトルの算出がしばしば行われる。ここでは、固有値計算や行列式計算、固有値分解や特異値分解等の線形代数の基本的な計算方法を示している。</b>

### 1) 固有値・固有ベクトルの求め方を確認する。

#### 1-1-1 スカラーとベクトルの違い

スカラー：ベクトル空間のベクトルに対比するものとしての実数。ベクトルの定数倍係数になれる。<br>
ベクトル：「大きさ」と「向き」を持った量。矢印で図示される。

#### 1-1-2 行列

・スカラーを表にしたもの<br>
・ベクトルを並べたもの<br>

$$
A=
\left[
    \begin{matrix} a_{11} & a_{12} & a_{13} \\ a_{21} & a_{22} & a_{23} \\ a_{31} & a_{32} & a_{33} \end{matrix}
\right]
$$

#### 1-1-3 行列とベクトルの積

2行2列の行列と2行1列のベクトルの積は以下のように計算される。

$$
\left[
\begin{matrix} 6 & 4 \\ 3 & 5 \end{matrix}
\right]
\left[
\begin{matrix} 1 \\ 2 \end{matrix}
\right]
=
\left[
\begin{matrix} 6*1 + 4*2 \\ 3*1 + 5*2 \end{matrix}
\right]
=
\left[
\begin{matrix} 6 + 8 \\ 3 + 10 \end{matrix}
\right]
=
\left[
\begin{matrix} 14 \\ 13 \end{matrix}
\right]
$$

#### 1-1-4 行列の積

3行3列同士の行列の積は以下のように定義される。

$$
\left[
\begin{matrix} a_{11} & a_{12} & a_{13} \\ a_{21} & a_{22} & a_{23} \\ a_{31} & a_{32} & a_{33} \end{matrix}
\right]
\left[
\begin{matrix} b_{11} & b_{12} & b_{13} \\ b_{21} & b_{22} & b_{23} \\ b_{31} & b_{32} & b_{33} \end{matrix}
\right]
=
\left[
\begin{matrix} a_{11}b_{11} + a_{12}b_{21} + a_{13}b_{31} &
               a_{11}b_{12} + a_{12}b_{22} + a_{13}b_{32} &
               a_{11}b_{13} + a_{12}b_{23} + a_{13}b_{33} \\
               a_{21}b_{11} + a_{22}b_{21} + a_{23}b_{31} &
               a_{21}b_{12} + a_{22}b_{22} + a_{23}b_{32} &
               a_{31}b_{13} + a_{32}b_{23} + a_{33}b_{33} \\
               a_{31}b_{11} + a_{32}b_{21} + a_{33}b_{31} & 
               a_{31}b_{12} + a_{32}b_{22} + a_{33}b_{32} &
               a_{31}b_{13} + a_{32}b_{23} + a_{33}b_{33} \end{matrix}
\right]
$$

＜例題＞

$$
    \left[
        \begin{matrix} 2 & 1 \\ 4 & 1 \end{matrix}
    \right]
    \left[
        \begin{matrix} 1 & 3 \\ 3 & 1 \end{matrix}
    \right]
= 
    \left[
        \begin{matrix} 2*1 + 1*3 & 2*3 + 1*1 \\ 4*1 + 1*3 & 4*3 + 1*1 \end{matrix}
    \right]
= 
    \left[
        \begin{matrix} 5 & 7 \\ 7 & 13 \end{matrix}
    \right]
$$

#### 1-1-5 連立1次方程式

$$
\begin{cases}
    x_{1} + 2x_{2} = 3 \\
    2x_{1} + 5x_{2} = 5
\end{cases}
$$

上の連立方程式を行列形式で表現すると以下のようになる。

$$
\left[
\begin{matrix} 1 & 2 \\ 2 & 5 \end{matrix}
\right]
\left[
\begin{matrix} x_{1} \\ x_{2} \end{matrix}
\right]
=
\left[
\begin{matrix} 3 \\ 5 \end{matrix}
\right]
$$

「[行列の基本変形](https://ja.wikipedia.org/wiki/%E8%A1%8C%E5%88%97%E3%81%AE%E5%9F%BA%E6%9C%AC%E5%A4%89%E5%BD%A2)」を用いて左辺の係数行列を単位行列になるまで変形することにより、右辺に連立方程式の解を得ることが出来る。

#### 1-1-6 逆行列

$$
A = 
\left[
\begin{matrix} a & b \\ c & d \end{matrix}
\right]
$$

上記行列Aにおいて、ad-bc≒0のとき、

$$
A^{-1} = 
\frac{1}{ad-bc}
\left[
\begin{matrix} a & b \\ c & d \end{matrix}
\right]
$$

をAの**逆行列**という。逆行列が定義出来るとき、Aを**正則である**、または**正則行列**と呼ぶ。逆行列は元の行列と掛けたときに**単位行列**となる性質がある。

$$
    AA^{-1} = A^{-1}A = I
$$

　(Iは単位行列)

逆行列は**[掃き出し法](https://ja.wikipedia.org/wiki/%E3%82%AC%E3%82%A6%E3%82%B9%E3%81%AE%E6%B6%88%E5%8E%BB%E6%B3%95)**（ガウスの消去法）により求めることが出来る。

#### 1-1-7 逆行列が存在しない条件

$$
A = 
\left[
\begin{matrix} a & b \\ c & d \end{matrix}
\right]
$$

において、ad-bc=0となるような行列は、逆行列が定義できない（正則ではない）。

$$
\left|
    \begin{matrix} a & b \\ c & d \end{matrix}
\right|
= ad - bc
$$

を行列Aの**行列式**といい、行列式が0か否かが逆行列の存在有無の判定条件となる。

$$
A=
\left[
    \begin{matrix} a_{11} & a_{12} & a_{13} \\ a_{21} & a_{22} & a_{23} \\ a_{31} & a_{32} & a_{33} \end{matrix}
\right]
$$

のような3行3列の行列の行列式は以下の通り。

$$
\left|
    \begin{matrix} a_{11} & a_{12} & a_{13} \\ a_{21} & a_{22} & a_{23} \\ a_{31} & a_{32} & a_{33} \end{matrix}
\right|
=
a_{11}
\left|
    \begin{matrix} a_{22} & a_{23} \\ a_{32} & a_{33} \end{matrix}
\right|
-a_{21}
\left|
    \begin{matrix} a_{12} & a_{13} \\ a_{32} & a_{33} \end{matrix}
\right|
+a_{31}
\left|
    \begin{matrix} a_{12} & a_{13} \\ a_{22} & a_{23} \end{matrix}
\right|
$$

＜例題1-1-7_1＞

以下の行列式を求めよ

$$
\left|
    \begin{matrix} 1 & 0 & -1 \\ 3 & 1 & 0 \\ 2 & -1 & 1 \end{matrix}
\right|
$$

＜解答1-1-7_1＞

$$
\left|
    \begin{matrix} 1 & 0 & -1 \\ 3 & 1 & 0 \\ 2 & -1 & 1 \end{matrix}
\right|
=
1
\left|
    \begin{matrix} 1 & 0 \\ -1 & 1 \end{matrix}
\right|
-3
\left|
    \begin{matrix} 0 & -1 \\ -1 & 1 \end{matrix}
\right|
+2
\left|
    \begin{matrix} 0 & -1 \\ 1 & 0 \end{matrix}
\right|
= 1*(1-0) - 3*(0-1) + 2*(0+1) = 6
$$

#### 1-1-8 固有値と固有ベクトル

ある行列Aに対して、以下の式が成り立つような、ベクトルxと係数λが存在する場合を考える。

$$
A\vec{x} = \lambda\vec{x}
$$

このようなベクトルを**固有ベクトル**といい、このときのLを**固有値**という。

＜例題1-1-8_1＞

$$
A = 
\left[
    \begin{matrix} 1 & 4 \\ 2 & 3 \end{matrix}
\right]
$$

$$
\vec{x} =
\left[
    \begin{matrix} 1 \\ 1 \end{matrix}
\right]
$$

の固有値を求めよ。

＜解答1-1-8_1＞

$$
\left[
    \begin{matrix} 1 & 4 \\ 1 & 1 \end{matrix}
\right]
\left[
    \begin{matrix} 1 \\ 1 \end{matrix}
\right]
=
\left[
    \begin{matrix} 5 \\ 5 \end{matrix}
\right]
= 5
\left[
    \begin{matrix} 1 \\ 1 \end{matrix}
\right]
$$

より、固有値（の内の1つ）は

$$
    \lambda = 5
$$

となる。

＜例題1-1-8_2＞

$$
A=
\left[
    \begin{matrix} 3 & 2 & 0 \\ 0 & 2 & 0 \\ 0 & 0 & 1 \end{matrix}
\right]
$$

の固有値、固有ベクトルを求めよ。

＜解答1-1-8_2＞

$$
A\vec{x} = \lambda\vec{x}
$$

の右辺を左辺に移行して

$$
(A - \lambda I)\vec{x} = \vec{0}
$$

となり、ここで

$$
\vec{x} \neq \vec{0}
$$

であるから、

$$
|A - \lambda I| = 0
$$

となる。この行列式を計算すると

$$
\left|
    \begin{matrix} 3-\lambda & 2 & 0 \\ 0 & 2-\lambda & 0 \\ 0 & 0 & 1-\lambda \end{matrix}
\right|
= (3-\lambda)(2-\lambda)(1-\lambda) = 0
$$

となるので、固有値λは

$$
\lambda = 3, 2, 1
$$

となる。

λ = 3のとき、

$$
\left[
    \begin{matrix} 3 & 2 & 0 \\ 0 & 2 & 0 \\ 0 & 0 & 1 \end{matrix}
\right]
\left[
    \begin{matrix} x_{1} \\ x_{2} \\ x_{3} \end{matrix}
\right]
=
3\left[
    \begin{matrix} x_{1} \\ x_{2} \\ x_{3} \end{matrix}
\right]
$$

となるので、固有ベクトルは

$$
\vec{x} =
k \left[
    \begin{matrix} 1 \\ 0 \\ 0 \end{matrix}
\right]
$$

λ = 2のとき、

$$
\left[
    \begin{matrix} 3 & 2 & 0 \\ 0 & 2 & 0 \\ 0 & 0 & 1 \end{matrix}
\right]
\left[
    \begin{matrix} x_{1} \\ x_{2} \\ x_{3} \end{matrix}
\right]
=
2\left[
    \begin{matrix} x_{1} \\ x_{2} \\ x_{3} \end{matrix}
\right]
$$

となるので、固有ベクトルは

$$
\vec{x} =
k \left[
    \begin{matrix} 2 \\ -1 \\ 0 \end{matrix}
\right]
$$

λ = 1のとき、

$$
\left[
    \begin{matrix} 3 & 2 & 0 \\ 0 & 2 & 0 \\ 0 & 0 & 1 \end{matrix}
\right]
\left[
    \begin{matrix} x_{1} \\ x_{2} \\ x_{3} \end{matrix}
\right]
=
\left[
    \begin{matrix} x_{1} \\ x_{2} \\ x_{3} \end{matrix}
\right]
$$

となるので、固有ベクトルは

$$
\vec{x} =
k \left[
    \begin{matrix} 0 \\ 0 \\ 1 \end{matrix}
\right]
$$

となる。ここで、kは比例定数である。

### 2) 固有値分解について理解を深める。

#### 1-2-1 固有値分解

[固有値分解](https://ja.wikipedia.org/wiki/%E5%9B%BA%E6%9C%89%E5%80%A4%E5%88%86%E8%A7%A3)とは、固有値に着目した行列の分解手法。ある実数を正方形に並べて作られた行列Aが固有値λ1, λ2, …と固有ベクトルv1, v2, …を持っていたとき、この固有値を対角線上に並べた行列（それ以外の成分は0）

$$
  \varLambda = \left[
    \begin{array}{cccc}
      \lambda_{1} & 0 & \ldots & 0 \\
      0 & \lambda_{2} & \ldots & 0 \\
      \vdots & \vdots & \ddots & \vdots \\
      0 & 0 & \ldots & \lambda_{n}
    \end{array}
  \right]
$$

とそれに対応する固有ベクトルを並べた行列

$$
  V = \left[
    \begin{array}{cccc}
      \vec{v}_{1} & \vec{v}_{2} & \ldots & \vec{v}_{n}
    \end{array}
  \right]
$$

を用意したとき、それらは

$$
    AV = V\varLambda 
$$

と関係づけられる。したがって

$$
    A = V\varLambda V^{-1} 
$$

と変形できる。このように正方形の行列を上述のような3つの行列の積に変換することを**固有値分解**という。この変換を用いることで累乗の計算が容易になる等の利点がある。

#### 1-2-2 固有値分解 具体例

＜具体例＞

$$
\left[
    \begin{matrix} 1 & 4 \\ 2 & 3 \end{matrix}
\right]
=
\left[
    \begin{matrix} 1 & 1 \\ 1 & -1/2 \end{matrix}
\right]
\left[
    \begin{matrix} 5 & 0 \\ 0 & -1 \end{matrix}
\right]
\left[
    \begin{matrix} 1/3 & 2/3 \\ 2/3 & -2/3 \end{matrix}
\right]
$$

＜例題＞

#### 1-2-3 固有値分解 問題

＜例題1-2-3_1＞

$$
\left[
    \begin{matrix} 2 & 1 \\ 0 & 6 \end{matrix}
\right]
$$

を固有値分解せよ。

＜解答1-2-3_1＞

標準形の行列式を解いて

$$
\left|
    \begin{matrix} 2-\lambda & 1 \\ 0 & 6-\lambda \end{matrix}
\right|
= (2-\lambda)(6-\lambda) = 0
$$

$$
    \lambda = 2, 6
$$

λ=2のとき

$$
\vec{x} =
k \left[
    \begin{matrix} 1 \\ 0 \end{matrix}
\right]
$$

λ=6のとき

$$
\vec{x} =
k \left[
    \begin{matrix} 1 \\ 4 \end{matrix}
\right]
$$

が固有ベクトルとなる（ただし、kは比例定数）。そのため、元の行列を固有値分解すると以下のようになる。

$$
\left[
    \begin{matrix} 2 & 1 \\ 0 & 6 \end{matrix}
\right]
=
\left[
    \begin{matrix} 1 & 1 \\ 0 & 4 \end{matrix}
\right]
\left[
    \begin{matrix} 2 & 0 \\ 0 & 6 \end{matrix}
\right]
\left[
    \begin{matrix} 1 & 1 \\ 0 & 4 \end{matrix}
\right]^{-1}
=
\left[
    \begin{matrix} 1 & 1 \\ 0 & 4 \end{matrix}
\right]
\left[
    \begin{matrix} 2 & 0 \\ 0 & 6 \end{matrix}
\right]
\left[
    \begin{matrix} 1 & -1/4 \\ 0 & 1/4 \end{matrix}
\right]
$$

### 3) 特異値・特異ベクトルの概要を知る。

#### 1-3-1 特異値分解

[特異値分解](https://ja.wikipedia.org/wiki/%E7%89%B9%E7%95%B0%E5%80%A4%E5%88%86%E8%A7%A3)は線形代数学における複素数あるいは実数を成分とする行列に対する行列分解の一手法。ある非正方行列Mについて

$$
M\vec{v} = \sigma\vec{u} \\
M^{T}\vec{u} = \sigma\vec{v}
$$

という特殊な単位ベクトルを定義できるとき、以下のように**特異値分解**できる。

$$
    M = USV^{-1} 
$$

### 4) 特異値分解の概要を知る。

#### 1-4-1 特異値の求め方

$$
MV = US \\
M^{T}U = CS^{T} \\
M = USV^{-1} \\
M^{T} = VS^{T}U^{-1}
$$

これらの積は

$$
MM^{T} = USV^{-1}VS^{T}U^{-1} = USS^{T}U^{-1}
$$

つまりMM^Tを固有値分解すれば、その左特異ベクトル（ただし単位ベクトルからつくらなければならないことに注意）と特異値の2乗が求められることが分かる。

#### 1-4-2 特異値の求め方 具体例

＜具体例1-4-2_1＞

$$
M = 
\left[
    \begin{matrix} 1 & 2 & 3 \\ 3 & 2 & 1 \end{matrix}
\right]
$$

を特異値分解せよ。

＜解答1-4-2_1＞

$$
MM^{T} = 
\left[
    \begin{matrix} 1 & 2 & 3 \\ 3 & 2 & 1 \end{matrix}
\right]
\left[
    \begin{matrix} 1 & 2 \\ 2 & 2 \\ 3 & 1 \end{matrix}
\right]
=
\left[
    \begin{matrix} 14 & 10 \\ 10 & 14 \end{matrix}
\right]
$$

であるから、これを固有値分解して

$$
\left[
    \begin{matrix} 14 & 10 \\ 10 & 14 \end{matrix}
\right]
=
\left[
    \begin{matrix} \frac{1}{\sqrt{2}} & -\frac{1}{\sqrt{2}} \\ \frac{1}{\sqrt{2}} & \frac{1}{\sqrt{2}} \end{matrix}
\right]
\left[
    \begin{matrix} 24 & 0 \\ 0 & 4 \end{matrix}
\right]
\left[
    \begin{matrix} \frac{1}{\sqrt{2}} & -\frac{1}{\sqrt{2}} \\ \frac{1}{\sqrt{2}} & \frac{1}{\sqrt{2}} \end{matrix}
\right]^{-1}
$$

となる。また、

$$
M^{T}M = 
\left[
    \begin{matrix} 1 & 2 \\ 2 & 2 \\ 3 & 1 \end{matrix}
\right]
\left[
    \begin{matrix} 1 & 2 & 3 \\ 3 & 2 & 1 \end{matrix}
\right]
=
\left[
    \begin{matrix} 10 & 8 & 6 \\ 8 & 8 & 8 \\ 6 & 8 & 10 \end{matrix}
\right]
$$

であるから、これを固有値分解して

$$
\left[
    \begin{matrix} 10 & 8 & 6 \\ 8 & 8 & 8 \\ 6 & 8 & 10 \end{matrix}
\right]
=
\left[
    \begin{matrix} \frac{1}{\sqrt{3}} & \frac{1}{\sqrt{2}} & \frac{1}{\sqrt{6}} \\ \frac{1}{\sqrt{3}} & 0 & -\frac{2}{\sqrt{6}} \\ \frac{1}{\sqrt{3}} & -\frac{1}{\sqrt{2}} & \frac{1}{\sqrt{6}} \end{matrix}
\right]
\left[
    \begin{matrix} 24 & 0 & 0 \\ 0 & 4 & 0 \\ 0 & 0 & 0 \end{matrix}
\right]
\left[
    \begin{matrix} \frac{1}{\sqrt{3}} & \frac{1}{\sqrt{2}} & \frac{1}{\sqrt{6}} \\ \frac{1}{\sqrt{3}} & 0 & -\frac{2}{\sqrt{6}} \\ \frac{1}{\sqrt{3}} & -\frac{1}{\sqrt{2}} & \frac{1}{\sqrt{6}} \end{matrix}
\right]^{-1}
$$

となるので、元の行列Mは以下のように特異値分解される。

$$
M = 
\left[
    \begin{matrix} 1 & 2 & 3 \\ 3 & 2 & 1 \end{matrix}
\right]
=
\left[
    \begin{matrix} \frac{1}{\sqrt{2}} & -\frac{1}{\sqrt{2}} \\ \frac{1}{\sqrt{2}} & \frac{1}{\sqrt{2}} \end{matrix}
\right]
\left[
    \begin{matrix} \frac{2}{\sqrt{6}} & 0 & 0 \\ 0 & 2 & 0 \end{matrix}
\right]
\left[
    \begin{matrix} \frac{1}{\sqrt{3}} & \frac{1}{\sqrt{3}} & \frac{1}{\sqrt{3}} \\ \frac{1}{\sqrt{2}} & 0 & -\frac{1}{\sqrt{2}} \\ \frac{1}{\sqrt{6}} & -\frac{2}{\sqrt{6}} & \frac{1}{\sqrt{6}} \end{matrix}
\right]
$$

特異値分解した画像データの行列から成分の小さい部分を取り除いていくとデータ量を小さくできる（低解像度化）。

## 第二章　確率・統計

<b>要約まとめ</b>

<b>与えられたデータの特性を把握し未知のデータに対して予測を立てることは、平均や分散などの基本統計量から母集団の確率分布を推定する処理である。実際には初めの段階でデータが十分に揃っておらず、後からデータが追加されてくる場合などもある。ここで示すベイズ則は、このような追加情報が得られ事前情報を新たな事後情報へと更新する場合などに用いられており、機械学習と親和性が高い。</b>

### 1) 条件付き確率について理解を深める。

#### 2-1-1 確率

・頻度確率（客観確率）・・・発生する頻度<br>
・ベイズ確率（主観確率）・・・信念の度合い

#### 2-1-2 条件付き確率

ある事象X=xが与えられたもとで、Y=yとなる事象

$$
P(Y=y|X=x) = 
\frac{P(Y=y, X=x)}{P(X=x)}
$$

#### 2-1-3 独立な事象の同時確率

お互いの発生には因果関係のない事象X=xとY=yが同時に発生する確率

$$
P(X=x, Y=y) = P(X=x)P(Y=y) = P(Y=y,X=x)
$$

### 2) ベイズ則の概要を知る。

#### 2-2-1 ベイズ則

ある街の子供たちは毎日1/4の確率で飴玉をもらうことができる（P(A)=1/4）<br>
飴玉をもらうと1/2の確率で笑顔になる（P(B|A)=1/2）<br>
笑顔の子供が飴玉をもらっている確率を求める。<br>
ただし、街の子供が笑顔でいる確率は1/3である（P(B)=1/3）。

P(A∧B) = P(B|A)×P(A) = 1/2×1/4 = 1/8<br>
P(A|B) = P(A∧B)/P(B) = (1/8) / (1/3) = 3/8

#### 2-2-2 確率変数と確率分布

確率変数・・・事象と結び付けられた数値。事象そのものをさすと解釈する場合も多い。<br>
確率分布・・・事象の発生する確率の分布。離散地であれば表に示せる。

### 3) 期待値・分散の求め方を確認する。

#### 2-3-1 期待値

その分布における、確率変数の平均の値 or 「ありえそう」な値

#### 2-3-2 分散と共分散

分散・・・データの散らばり具合。データの各々の値が、期待値からどれだけずれているのか平均したもの。<br>
共分散・・・2つのデータ系列の傾向の違い。正の値を取れば似た傾向、負の値を取れば逆の傾向、0を取れば関係性に乏しい。

#### 2-3-3 分散と標準偏差

分散は2乗してしまっているので元のデータと単位が違う→平方根を取ることによって元の単位に戻す

### 4) 様々な確率分布の概要を知る。

#### 2-4-1 様々な確率分布Ⅰ

ベルヌーイ分布・・・二者択一の確率事象を表す分布（コイントス等）<br>
マルチヌーイ（カテゴリカル）分布・・・サイコロを転がすイメージ

#### 2-4-2 様々な確率分布Ⅱ

二項分布・・・ベルヌーイ分布の多試行版<br>
ガウス分布・・・釣鐘型の連続分布

## 第三章　情報理論

<b>要点まとめ</b>

<b>高確率で発生する出来事は情報量が低く、稀にしか発生しない出来事は情報量が高いとされる（例えば、さいころを複数回振った時の「1以外が続けて出る」事象よりも「1が連続して出る」事象の方がレアであり、情報量としては高い）。このような、確率 𝑝 で生起する事象が起きたことを知ったときに得られる情報量 I(p) を「自己情報量」といい、この自己情報量の期待値を「シャノンエントロピー」という。また、2つの確率分布間の再を計る尺度として用いられるのが「KLダイバージェンス」であり、これは「交差エントロピー」と「エントロピー」の差で求められる。</b>

### 1) 自己情報量・シャノンエントロピーの定義を確認する。

#### 3-1-1 自己情報量

・対数の底が2のとき、単位はビット（bit）<br>
・対数の底がネイピア数eのとき、単位はナット（nat）

$$
I(x) = -log(P(x)) = log(W(x))
$$

#### 3-1-2 シャノンエントロピー

自己情報量の期待値

$$
H(x) = E(I(x)) = -E(log(P(x))) = -\sum(P(x)log(P(x)))
$$

### 2) KLダイバージェンス・交差エントロピーの概要を知る。

#### 3-2-1 カルバック・ライブラー ダイバージェンス

同じ事象・確率変数における異なる確率分布P, Qの違いを表す。

#### 3-2-2 交差エントロピー

KLダイバージェンスの一部分を取り出したもの<br>
Qについての自己情報量をPの分布で平均している。

<div style="text-align: center;">
################################################################################################
</div>

## 演習問題

### 第1章　ベクトルと行列の演算Ⅰ

#### 1.1

次のベクトル
$$
\vec{a} =
\left[
    \begin{matrix} 1 \\ 6 \\ 3 \end{matrix}
\right]
  , 
\vec{b} =
\left[
    \begin{matrix} 5 \\ 2 \\ 4 \end{matrix}
\right]
$$
に関して、以下の計算をせよ。

問 1.1.1 $$ \vec{a} + \vec{b} $$

$$
\vec{a} + \vec{b} =
\left[
    \begin{matrix} 1 \\ 6 \\ 3 \end{matrix}
\right]
+
\left[
    \begin{matrix} 5 \\ 2 \\ 4 \end{matrix}
\right]
=
\left[
    \begin{matrix} 6 \\ 8 \\ 7 \end{matrix}
\right]
$$

In [2]:
import numpy as np
a = np.array([1, 6, 3])
b = np.array([5, 2, 4])
print(a + b)

[6 8 7]


問 1.1.2 $$ \vec{a} - \vec{b} $$

$$
\vec{a} + \vec{b} =
\left[
    \begin{matrix} 1 \\ 6 \\ 3 \end{matrix}
\right]
-
\left[
    \begin{matrix} 5 \\ 2 \\ 4 \end{matrix}
\right]
=
\left[
    \begin{matrix} -4 \\ 4 \\ -1 \end{matrix}
\right]
$$

In [3]:
print(a - b)

[-4  4 -1]


問 1.1.3 $$ 7*\vec{a} $$

$$
7 * \vec{a} =
7 * \left[
    \begin{matrix} 1 \\ 6 \\ 3 \end{matrix}
\right]
=
\left[
    \begin{matrix} 7 \\ 42 \\ 21 \end{matrix}
\right]
$$

In [4]:
print(7*a)

[ 7 42 21]


問 1.1.4 $$ 8*(\vec{a} + \vec{b}) $$

$$
8 * (\vec{a} + \vec{b}) =
8 * \left[
    \begin{matrix} 6 \\ 8 \\ 7 \end{matrix}
\right]
=
\left[
    \begin{matrix} 48 \\ 64 \\ 56 \end{matrix}
\right]
$$

In [5]:
print(8*(a+b))

[48 64 56]


#### 1.2

次の行列
$$
A = 
\left[
    \begin{matrix} 2 & 1 \\ 5 & 3 \end{matrix}
\right]
  , 
B = 
\left[
    \begin{matrix} 1 & 4 \\ 1 & 5 \end{matrix}
\right]
$$
に関して、以下の計算をせよ。

問 1.2.1

$$ A + B $$

$$
A + B = 
\left[
    \begin{matrix} 2 & 1 \\ 5 & 3 \end{matrix}
\right]
+ 
\left[
    \begin{matrix} 1 & 4 \\ 1 & 5 \end{matrix}
\right]
=
\left[
    \begin{matrix} 3 & 5 \\ 6 & 8 \end{matrix}
\right]
$$

In [8]:
import numpy as np
A = np.array([[2, 1],[5, 3]])
B = np.array([[1, 4],[1, 5]])
print(A + B)

[[3 5]
 [6 8]]


問 1.2.2

$$ A - 3B $$

$$
A - 3*B = 
\left[
    \begin{matrix} 2 & 1 \\ 5 & 3 \end{matrix}
\right]
- 3* 
\left[
    \begin{matrix} 1 & 4 \\ 1 & 5 \end{matrix}
\right]
=
\left[
    \begin{matrix} -1 & -11 \\ 2 & -12 \end{matrix}
\right]
$$

In [9]:
print(A - 3*B)

[[ -1 -11]
 [  2 -12]]


### 第2章 ベクトルと行列の演算Ⅱ

#### 2.1

次のベクトルと行列
$$
\vec{v} =
\left[
    \begin{matrix} 1 \\ 0 \\ 3 \end{matrix}
\right]
  , 
A = 
\left[
    \begin{matrix} 1 & 3 & 4 \\ 5 & 9 & 0 \\ 3 & 1 & 2 \end{matrix}
\right]
  , 
B = 
\left[
    \begin{matrix} 1 & 0 & 3 \\ 0 & 2 & 5 \end{matrix}
\right]
$$
に関して、以下の計算をせよ。

問 2.1.1

$$ A * \vec{v} $$

$$ A * \vec{v} = 
\left[
    \begin{matrix} 1 & 3 & 4 \\ 5 & 9 & 0 \\ 3 & 1 & 2 \end{matrix}
\right]
\left[
    \begin{matrix} 1 \\ 0 \\ 3 \end{matrix}
\right]
=
\left[
    \begin{matrix} 13 \\ 5 \\ 9 \end{matrix}
\right]
$$

In [20]:
import numpy as np
v = np.array([[1], [0], [3]])
A = np.array([[1, 3, 4],[5, 9, 0],[3, 1, 2]])
B = np.array([[1, 0, 3],[0, 2, 5]])
print(A @ v)

[[13]
 [ 5]
 [ 9]]


問 2.1.2

$$ B * \vec{v} $$

$$ B * \vec{v} = 
\left[
    \begin{matrix} 1 & 0 & 3 \\ 0 & 2 & 5 \end{matrix}
\right]
\left[
    \begin{matrix} 1 \\ 0 \\ 3 \end{matrix}
\right]
=
\left[
    \begin{matrix} 10 \\ 15 \end{matrix}
\right]
$$

In [21]:
print(B @ v)

[[10]
 [15]]


問 2.1.3

$$ BA $$

$$ BA = 
\left[
    \begin{matrix} 1 & 0 & 3 \\ 0 & 2 & 5 \end{matrix}
\right]
\left[
    \begin{matrix} 1 & 3 & 4 \\ 5 & 9 & 0 \\ 3 & 1 & 2 \end{matrix}
\right]
=
\left[
    \begin{matrix} 10 & 6 & 10 \\ 25 & 23 & 10 \end{matrix}
\right]
$$

In [22]:
print(B @ A)

[[10  6 10]
 [25 23 10]]


問 2.1.4

$$ B^{T} $$

$$
B^{T} = 
\left[
    \begin{matrix} 1 & 0 \\ 0 & 2 \\ 3 & 5 \end{matrix}
\right]
$$

In [23]:
print(B.T)

[[1 0]
 [0 2]
 [3 5]]


#### 2.2

次の行列
$$
A = 
\left[
    \begin{matrix} 2 & 1 \\ 4 & 1 \end{matrix}
\right]
  , 
B = 
\left[
    \begin{matrix} 1 & 3 \\ 3 & 1 \end{matrix}
\right]
$$
に関して、以下の計算をせよ。

問 2.2.1

$$ AB $$

$$
AB = 
\left[
    \begin{matrix} 2 & 1 \\ 4 & 1 \end{matrix}
\right]
\left[
    \begin{matrix} 1 & 3 \\ 3 & 1 \end{matrix}
\right]
=
\left[
    \begin{matrix} 5 & 7 \\ 7 & 13 \end{matrix}
\right]
$$

In [24]:
import numpy as np
A = np.array([[2, 1],[4, 1]])
B = np.array([[1, 3],[3, 1]])
print(A @ B)

[[ 5  7]
 [ 7 13]]


問 2.2.2

$$ A^{-1} $$

$$
A^{-1} = 
\left[
    \begin{matrix} -1/2 & 1/2 \\ 2 & -1 \end{matrix}
\right]
$$

In [25]:
print(np.linalg.inv(A))

[[-0.5  0.5]
 [ 2.  -1. ]]


問 2.2.3

$$ B^{-1} $$

$$
B^{-1} = 
\left[
    \begin{matrix} -1/8 & 3/8 \\ 3/8 & -1/8 \end{matrix}
\right]
$$

In [26]:
print(np.linalg.inv(B))

[[-0.125  0.375]
 [ 0.375 -0.125]]


問 2.2.4

$$ BAB^{-1} $$

$$
BAB^{-1} = 
\left[
    \begin{matrix} 1 & 3 \\ 3 & 1 \end{matrix}
\right]
\left[
    \begin{matrix} 2 & 1 \\ 4 & 1 \end{matrix}
\right]
\left[
    \begin{matrix} -1/8 & 3/8 \\ 3/8 & -1/8 \end{matrix}
\right]
=
\left[
    \begin{matrix} -1/4 & 4+3/4 \\ 1/4 & 3+1/4 \end{matrix}
\right]
$$

In [27]:
print(B @ A @ np.linalg.inv(B))

[[-0.25  4.75]
 [ 0.25  3.25]]


### 第3章 確率変数と確率分布

問 3.1

試行の結果として起こる事象に整数や実数の数値が結び付けられているときに、その数値を「確率変数」という。<br>
次の選択肢のうち、確率変数として適当なものはどれか。すべて選べ。

a. さいころを振ったときに出た目の数<br>
b. 3枚のコインを同時に投げて裏表どちらが出るか試行した時のコインの枚数<br>
c. 赤, 青, 黄の3色のボールが壺の中に入っている。個々から1個を取り出した時の色<br>
d. 8本のうち1本が当たりであるくじを、当たりが出るまで抽選し続けた時の回数

→aとd<br>
bは試行結果と無関係なので確率変数ではない<br>
cは数値と結びついていないので確率変数ではない

問 3.2

4枚のコインを同時に投げる試行を1200回行ったとして、下表を埋めよ。

| 事象 | 裏が0枚,<br>表が4枚 | 裏が1枚,<br>表が3枚 | 裏が2枚,<br>表が2枚 | 裏が3枚,<br>表が1枚 | 裏が4枚,<br>表が0枚 |
| ---- | ---- | ---- | ---- | ---- | ---- |
| 確率変数（裏を0,<br>表を1と対応させ<br>和を取った） | 4 | 3 | 2 | 1 | 0 |
| 事象が発生した回数 | 75 | 300 | 450 | 300 | 75 |
| 事象と対応する確率 | 1/16 | 4/16 | 6/16 | 4/16 | 1/16 |

### 第4章 情報量

4.1

問 4.1.1　1枚のコインを1回投げて表が出たという事象の情報量は何bitか。

$$
I = -log_{2}(p(X)) = -log_{2}(1/2) = 1(bit)
$$

問 4.1.2　2枚のコインを1回投げてすべて表が出たという事象の情報量は何bitか。

$$
I = -log_{2}(p(X)) = -log_{2}(1/2 * 1/2) = 2(bit)
$$

問 4.1.3　n枚のコインを1回投げて1枚の表が出たという事象の情報量は何bitか。

$$
I = -log_{2}(p(X)) = -log_{2}(n * (1/2)^{n}) = -log_{2}(n) + n(bit)
$$

### 第5章 条件付き確率

問 5.1

1年のうち洗濯物を干していた日数を60日、洗濯物を干していてかつ雨が降ってきた日数を12日として、ある年における、洗濯物を干していた日という条件下で、雨が降ってきた日の発生する確率と、洗濯物を干していてかつ雨が降ってきた日の発生する確率を求めよ。

事象A：洗濯物を干す<br>
事象B：雨が降る

$$ P(A) = \frac{60}{365} $$
$$ P(A \cap B) = \frac{12}{365} $$
$$ P(A | B) = \frac{P(A \cap B)}{P(A)} = \frac{\frac{12}{365}}{\frac{60}{365}} = \frac{1}{5}$$

問 5.2

袋の中に赤い玉3個と白い玉2個が入っている。赤い玉はA, B, Cの文字が、白い玉にはA, Bの文字が、それぞれ1個に対して1文字ずつ記されている。以下の問いに答えよ。

問 5.2.1

出てきた玉が赤色であったとき、それに記されている文字がBである確率。

事象A：出てきた玉が赤色である。<br>
事象B：出てきた玉に記されている文字がB。

$$ P(A) = \frac{3}{5} $$
$$ P(B) = \frac{2}{5} $$
$$ P(A \cap B ) = \frac{1}{5} $$
$$ P(A | B) = \frac{P(A \cap B )}{P(A)} = \frac{\frac{1}{5}}{\frac{3}{5}} = \frac{1}{3} $$

問 5.2.2

出てきた玉に記されている文字がAであったとき、その玉の色が白色である確率。

事象A：出てきた玉に記されている文字がA。<br>
事象B：出てきた玉が白色。

$$ P(A) = \frac{2}{5} $$
$$ P(B) = \frac{2}{5} $$
$$ P(A \cap B) = \frac{1}{5} $$
$$ P(A | B) = \frac{P(A \cap B)}{P(A)} = \frac{\frac{1}{5}}{\frac{2}{5}} = \frac{1}{2}$$

### 第6章 対数と乗算除算の関係

問 6.1

実数X, A, Bに対して X = AB という関係が成り立っているとき、

$$ \log (X) = \log (AB) = \log (A) + \log (B) $$

問 6.2

X = A/B のとき、

$$ \log (X) = \log (\frac{A}{B}) = \log (A) - \log (B) $$

問 6.3

実数X, x1, x2, x3, x4に対して X=x1x2x3x4 という関係が成り立っているとき、

$$ \log(X) = \log(x_{1}x_{2}x_{3}x_{4}) = \log(x_{1}) + \log(x_{2}) + \log(x_{3}) + \log(x_{4}) = \sum_{i=1}^{4}\log(x_{i})$$

### 第7章 確認テスト

問 7.1

次のベクトルの和を求めよ。

$$
\vec{a} + \vec{b} =
\left[
    \begin{matrix} 2 \\ 6 \\ 3 \end{matrix}
\right]
+
\left[
    \begin{matrix} 1 \\ 1 \\ 4 \end{matrix}
\right]
=
\left[
    \begin{matrix} 3 \\ 7 \\ 7 \end{matrix}
\right]
$$

問 7.2

行列(1 4;2 3)の固有ベクトル(1; 1)に対応する固有値を求めよ。

$$
\left[
    \begin{matrix} 1 & 4 \\ 2 & 3 \end{matrix}
\right]
\left[
    \begin{matrix} 1 \\ 1 \end{matrix}
\right]
= 5 *
\left[
    \begin{matrix} 1 \\ 1 \end{matrix}
\right]
$$

$$ \lambda = 5 $$

問 7.3

離散的な確率分布P(x)の下でのある確率変数f(x)の期待値

$$ E(f) = \Sigma P(x)f(x) $$

問 7.4

ある確率変数f(x)の分散

$$ Var(f) = E(( f(x) - E(f(x)) )^{2}) = E(f(x)^{2}) - E(f(x))^{2} $$

問 7.5

ある離散的な事象の確率分布をP(x)としたときのシャノンエントロピー

$$ -\Sigma P(x)\log(P(x)) $$